### ISSUE TITLE
Config parse error when using Extensible Identity Header and `2022-blake3-chacha20-poly1305` algorithm

### ISSUE THREAD

**Issue by bindreams** [shadowsocks/shadowsocks-rust#1525]
Here is an example config which should work with `sslocal`. It's using a the new encryption algorithm with EIH. According to the README, I should concatenate the server token and user token with `:`
```json
{
    "local_address": "127.0.0.1",
    "local_port": 1083,

    "servers": [
        {
            "server": "1.2.3.4",
            "server_port": 20000,
            "password": "lIeOeQwb9OXkAi7EWNhm62acn8rs97t0+mVdms7V6fE=:wjxYGa701LiKfdBD46U4yd0+rRU627V6ZB4rM5dzpwE=",
            "method": "2022-blake3-chacha20-poly1305"
        }
    ]
}
```

Instead of starting, the program panics when parsing the config:
```
$ bin/sslocal -v -c .\shadowsocks.json
thread 'main' panicked at D:\a\shadowsocks-rust\shadowsocks-rust\crates\shadowsocks\src\config.rs:364:17:
2022-blake3-chacha20-poly1305 password lIeOeQwb9OXkAi7EWNhm62acn8rs97t0+mVdms7V6fE=:wjxYGa701LiKfdBD46U4yd0+rRU627V6ZB4rM5dzpwE= is not base64 encoded, error: Invalid symbol 61, offset 43.
note: run with RUST_BACKTRACE=1 environment variable to display a backtrace
```

However, when switching the algorithm to `2022-blake3-aes-256-gcm`, everything gets parsed OK and the program starts.

My sslocal version is 1.18.4.

**Comment by zonyitoo** [comment]
Password of SIP022 ciphers are mandatory to be in a fixed length that the cipher defines.

Read the doc: https://shadowsocks.org/doc/sip022.html

**Comment by zonyitoo** [comment]
EIH is only supported by the AES ciphers. https://shadowsocks.org/doc/sip023.html

**Comment by bindreams** [comment]
Sorry... this just multiplies the number of questions that I have. Why does `ssserver` accept the "users" array with this cipher selected? Furthermore, when using a third-party local client (I use shadowrocket on iOS), the combination of cipher and EIH actually _works_, I have an internet connection and everything.

Please don't take this the wrong way, obviously I'm not an expert, but the SIP you linked to does not say that EIH is only supported by the AES ciphers. It says this:
```
Identity headers are implemented in such a way that's fully backward compatible with current
[Shadowsocks 2022](https://shadowsocks.org/doc/sip022.html) implementations.
```

And in the "Shadowsocks 2022 implementations" SIP it says the following:
```
4. Optional Methods
Implementations MAY choose to implement 2022-blake3-chacha20-poly1305,
2022-blake3-chacha12-poly1305 and 2022-blake3-chacha8-poly1305 when support
for CPUs without AES instructions is a priority.
```

So by my understanding, `2022-blake3-chacha20-poly1305` is a Shadowsocks 2022 implementation and should be compatible with EIH, no? If third-party implementations have successfully enabled it, would you consider this as a feature request?

**Comment by zonyitoo** [comment]
I must be wrong. Your error message indicated that the error was reported while parsing the configuration.

**Comment by zonyitoo** [comment]
Well yes, in this project, only AES ciphers are supported:

https://github.com/shadowsocks/shadowsocks-rust/blob/222adb38b3edb0100a1d5d723c33f4ff91010d02/crates/shadowsocks/src/config.rs#L390-L400

**Comment by zonyitoo** [comment]
> If third-party implementations have successfully enabled it, would you consider this as a feature request?

The EIH specification only mentioned how to implement it with those AES ciphers, how are those other implementations make the chacha20-* method works with EIH?

**Comment by zonyitoo** [comment]
I cannot find any documents about how to implement EIH with `2022-blake3-chacha20-poly1305`. @database64128 did you make any new extension for the EIH standard?

**Comment by bindreams** [comment]
I tried searching, but the iOS app is closed-source. I only know that however it encoded the data, `ssserver` from this project successfully decoded it. Maybe you could work backwards from the decode implementation.

**Comment by zonyitoo** [comment]
https://github.com/shadowsocks/shadowsocks-rust/blob/222adb38b3edb0100a1d5d723c33f4ff91010d02/crates/shadowsocks/src/relay/tcprelay/aead_2022.rs#L279-L281

As we just saw, `method_support_eih` must be `false` for `2022-blake3-chacha20-poly1305`. So in this project, it will completely skip the EIH decode block.

You can enable trace log by `-vvv` and see if you can find any logs like this:

https://github.com/shadowsocks/shadowsocks-rust/blob/222adb38b3edb0100a1d5d723c33f4ff91010d02/crates/shadowsocks/src/relay/tcprelay/aead_2022.rs#L329-L333

If `ssserver` can work for your client, I would make an assumption: your client didn't send EIH for `2022-blake3-chacha20-poly1305`.

**Comment by bindreams** [comment]
You are correct, I'm not seeing these messages with `2022-blake3-chacha20-poly1305` but I am seeing them with `2022-blake3-aes-256-gcm`. I also tried deleting all the users in the server config and the third-party app connected anyway.

I guess this is truly a feature request, if it is even possible to implement. If this cannot be implemented, I would suggest instead making stricter config checks on the server side, so that `ssserver` does not accepts `users` with unsupported ciphers.

I also just found something related to this, which is probably a security vulnerability. Is it ok to post it here?
